analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -40,20 +71,41 @@ public sealed record SymbolicValue
         public override string ToString() =>
             SerializeConstraints();
 
-        public SymbolicValue WithConstraint(SymbolicConstraint constraint) =>
-            HasConstraint(constraint)
-                ? this
-                : this with { Constraints = Constraints.SetItem(constraint.GetType(), constraint) };
+        public SymbolicValue WithConstraint(SymbolicConstraint constraint)
```

Comment:
What about this? It only changes the syntax, while keeping the logic. I find the original logic quite scattered and hard to follow.
The ideas:
* using if/elseif/else to clearly express the main decision points
* `constraint.GetType()` is fast enough and we don't need to extract it
* using switch statement to streamline the `1, 2, 3+` cases - this part still unfortunately bit irregular, but I don't see a better way. Maybe you'll spot something.
```
        public SymbolicValue WithConstraint(SymbolicConstraint constraint)
        {
            if (Constraints.Count == 0)
            {
                return SingleConstraint(constraint);
            }
            else if (HasConstraint(constraint))
            {
                return this;
            }
            else
            {
                var containsContraintType = Constraints.ContainsKey(constraint.GetType());
                return Constraints.Count switch
                {
                    1 when containsContraintType => SingleConstraint(constraint),
                    1 => PairConstraint(Constraints.Values.First(), constraint),
                    2 when containsContraintType => PairConstraint(OtherSingle(constraint.GetType()), constraint),
                    _ => this with { Constraints = Constraints.SetItem(constraint.GetType(), constraint) }
                };
            }
        }
```

Reply:
There was also the `else` part to make the overall flow explicit